### ISSUE TITLE
Receiving large messages

### ISSUE THREAD

**Issue by azku** [emqx/emqtt#46]
When receiving messages with few kilobytes, I found that memory usage shots up to unsustainable levels. Looking deeper into the code, I believe, the (very elegant) way in which packages are pulled together appending binaries and also returning functions with closures, could be pointed as being the "offender".
Would it be reasonable to change this to using iolist and perhaps avoid the return of functions that accumulate closures? Or am I misusing emqttc altogether?

**Comment by macpie** [comment]
Hey @azku,

I have encountered the same issue. My app is consuming few kb messages (thousands/sec). As soon as it would start I would see a memory increase that would eventually crash the app.

After some digging I realized that the back up was not in the [socket/parser](https://github.com/emqtt/emqttc#design)  but the actual `client fsm`. It would start stacking messages in the mialbox right away causing the "memory leak". 

It seems that the client is trying to [log every packet a debug level](https://github.com/emqtt/emqttc/blob/master/src/emqttc.erl#L639) 
```
connected(Packet = ?PACKET(_Type), State = #state{name = Name, logger = Logger}) ->
    Logger:debug("[Client ~s] RECV: ~s", [Name, emqttc_packet:dump(Packet)]),
    {ok, NewState} = received(Packet, State),
    next_state(connected, NewState);
```
which should be ok if you log level is higher that that problem is `emqttc_packet:dump(Packet)` is evaluated no matter what.
This  function seems to do a bunch of "string operation" on the packet to display it that are ending up being really costly.
By removing or commenting that line out, I observed a massive improvement, no more messages in the mailbox.

I will be issuing a pull req to comment this out for now.
